### DISCUSSION TITLE
reduce closing brackets on own line

### DISCUSSION THREAD

**Discussion by raylu** [jsh9/cercis#29] [Ideas]
something that has always annoyed me about black is it will place closing `)}]` on its own line
```
$ cercis --line-length 30 --code "print('hiiiiiiiiiiiiiiiiiiiiiiii')"
print(
    'hiiiiiiiiiiiiiiiiiiiiiiii'
)
```
```
$ cercis --line-length 30 --code "print('hiiiiiiiiii', 'thereeeeee')"
print(
    'hiiiiiiiiii',
    'thereeeeee',
)
```
I think this behavior is tolerable for `]` and `}`, but I really dislike it for function calls (and I rarely building giant tuples). my ideal formatting would be
```py
print(
    'hiiiiiiiiiiiiiiiiiiiii')  # this line is 29 chars long
print(
    'hiiiiiiiiiiiiiiiiiiiiiiii')  # this line is 32 chars long
print('hiiiiiiiiii',
    'thereeeeee')
```

if you could get me anywhere closer to that (especially for function calls in compound statements (`if`/`for`)), that'd be great. do you have any interest in supporting this?

**Comment by jsh9** [comment]
Hi @raylu ,

I implemented a hacky solution, which adds "fixes" on top of `Black`'s formatting result.  Please see the code changes here: https://github.com/jsh9/cercis/pull/30

This code change currently does **not** cover quite a few edge cases, such as this:

```python
myFunc(
    arg1,
    arg2,
    arg3,
    myFunc2(
        arg4,
        arg5,
        arg6),
    arg7)
```

And it is this example above that made me think that collapsing closing parenthesis may not be the best idea, because it can sort of lead to a "messy" rabbit hole.

If I am already collapsing `)` behind `arg6`, why don't I collapse `arg7` to the previous line?  And if I'm already collapsing function calls, shall I also collapse function definitions then?

If you really feel strongly about implementing this style, please feel free to take over this PR of mine and add more test cases.  Otherwise, I'm inclined to not support this.

**Comment by raylu** [comment]
hm... I'm starting to see why yapf tries to search the space of possible formats for the least cost now. the result is inscrutable, but you can get nice results like
```py
outer_call(arg1, arg2, arg3,
    inner_call(arg4, arg5, arg6), arg7)
```

I don't feel strongly. thanks for the quick response!
